File: backend/app/api/mdoc.py
from os import cpu_count
from fastapi import APIRouter, HTTPException
from pathlib import Path
from typing import List, Dict
import shutil
import asyncio

from app.models.types import (
    TiltSeries,
    ScanConfig,
    MdocScanResponse,
    BatchSaveRequest,
    BatchSaveResponse,
    BackupDeleteRequest,
    BackupDeleteResponse,
)
from pydantic import BaseModel
from app.state.project_state import project_state
from app.matcher.cut_match import ImageMatcher
from app.mdoc.parser import parse_mdoc_file
from app.mdoc.writer import write_mdoc_with_selections

router = APIRouter()


# ===== SAVE ALL REQUEST/RESPONSE =====

class SaveAllRequest(BaseModel):
    """Request to save all mdoc changes. Frontend sends all selections."""
    selections: Dict[str, Dict[int, bool]]  # mdocPath -> {zIndex: selected}


class SaveAllResponse(BaseModel):
    """Response from save all operation."""
    success: bool
    saved: List[str] = []
    failed: List[str] = []
    deleted: List[str] = []
    message: str


class DeleteAllRequest(BaseModel):
    """Request to delete multiple mdoc files."""
    mdocPaths: List[str]


@router.post("/scan", response_model=MdocScanResponse)
async def scan_project(config: ScanConfig):
    """Scan project directory for mdoc files"""
    try:
        project_state.set_config(config)

        # Build image matcher cache
        matcher = ImageMatcher(
            config.image_dir,
            config.image_prefix_cut,
            config.image_suffix_cut,
        )
        matcher.build_cache()

        # Scan for mdoc files
        mdoc_dir = Path(config.mdoc_dir)
        if not mdoc_dir.exists():
            raise HTTPException(
                status_code=404, detail=f"mdoc directory not found: {config.mdoc_dir}"
            )

        # Collect all mdoc files first
        mdoc_files = list(mdoc_dir.rglob("*.mdoc"))
        print(f"Found {len(mdoc_files)} mdoc files to scan")

        # Parse mdoc files in parallel
        from concurrent.futures import ThreadPoolExecutor

        def parse_single_mdoc(mdoc_file_path: str):
            """Parse a single mdoc file - runs in worker thread"""
            try:
                ts = parse_mdoc_file(mdoc_file_path, matcher)
                return ts
            except Exception as e:
                print(f"Warning: Failed to parse {mdoc_file_path}: {e}")
                return None

        # Use ThreadPoolExecutor for parallel parsing (can share matcher object)
        tilt_series: List[TiltSeries] = []
        max_workers = min(cpu_count() or 4, len(mdoc_files))

        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            # Submit all parsing tasks
            future_to_file = {
                executor.submit(parse_single_mdoc, str(mdoc_file)): mdoc_file
                for mdoc_file in mdoc_files
            }

            # Collect results as they complete
            for future in future_to_file:
                try:
                    ts = await asyncio.get_event_loop().run_in_executor(
                        None, future.result
                    )
                    if ts:
                        project_state.add_tilt_series(ts)
                        tilt_series.append(ts)
                except Exception as e:
                    mdoc_file = future_to_file[future]
                    print(f"Warning: Failed to parse {mdoc_file}: {e}")

        print(f"Successfully scanned {len(tilt_series)} tilt series")

        return MdocScanResponse(tiltSeries=tilt_series, total=len(tilt_series))

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Scan failed: {str(e)}")


@router.get("/list")
async def list_tilt_series():
    """List all tilt series"""
    return project_state.list_tilt_series()


@router.get("/{ts_id}")
async def get_tilt_series(ts_id: str):
    """Get specific tilt series"""
    ts = project_state.get_tilt_series(ts_id)
    if not ts:
        raise HTTPException(status_code=404, detail=f"Tilt series not found: {ts_id}")
    return ts


@router.post("/batch-save", response_model=BatchSaveResponse)
async def batch_save(request: BatchSaveRequest):
    """
    Save single mdoc file directly to disk.
    NOTE: Consider using /save-all for bulk operations.
    """
    try:
        # Find tilt series
        ts = next((t for t in project_state.list_tilt_series() if t.mdocPath == request.mdocPath), None)
        if not ts:
            raise HTTPException(status_code=404, detail=f"Tilt series not found: {request.mdocPath}")

        # Write directly to disk
        backup_path = write_mdoc_with_selections(request.mdocPath, request.selections)
        
        # Update in-memory state
        project_state.update_tilt_series_frames(request.mdocPath, request.selections)

        return BatchSaveResponse(
            success=True,
            message=f"Saved {len(request.selections)} frame selections",
            backupPath=backup_path,
            updatedTiltSeries=project_state.get_tilt_series(ts.id),
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=409, detail=f"Save failed: {str(e)}")


@router.post("/backup-delete", response_model=BackupDeleteResponse)
async def backup_delete(request: BackupDeleteRequest):
    """Backup and delete single mdoc file"""
    try:
        mdoc_path = Path(request.mdocPath)
        if not mdoc_path.exists():
            raise HTTPException(
                status_code=404, detail=f"mdoc file not found: {request.mdocPath}"
            )

        # Create backup
        backup = mdoc_path.with_suffix(".mdoc.bak")
        shutil.copy2(mdoc_path, backup)

        # Delete original
        mdoc_path.unlink()

        # Remove from project state
        project_state.remove_tilt_series_by_mdoc_path(request.mdocPath)

        return BackupDeleteResponse(
            success=True,
            message=f"Backed up and deleted {request.mdocPath}",
            backupPath=str(backup),
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=409, detail=f"Backup-delete failed: {str(e)}")


# ===== UNIFIED SAVE ALL API =====

@router.post("/save-all", response_model=SaveAllResponse)
async def save_all(request: SaveAllRequest):
    """
    Save all mdoc changes in one request.
    Frontend sends all selections: {mdocPath: {zIndex: selected}}
    Backend writes directly to disk, no staging.
    """
    saved = []
    failed = []
    
    if not request.selections:
        return SaveAllResponse(
            success=True,
            message="No changes to save",
            saved=[],
            failed=[],
            deleted=[]
        )
    
    # Write each mdoc file directly to disk
    for mdoc_path, selections in request.selections.items():
        try:
            # Verify tilt series exists
            ts = next((t for t in project_state.list_tilt_series() if t.mdocPath == mdoc_path), None)
            if not ts:
                failed.append(f"{mdoc_path}: tilt series not found")
                continue
            
            # Write to disk
            write_mdoc_with_selections(mdoc_path, selections)
            
            # Update in-memory state
            project_state.update_tilt_series_frames(mdoc_path, selections)
            
            saved.append(mdoc_path)
            
        except Exception as e:
            print(f"Failed to save {mdoc_path}: {e}")
            failed.append(f"{mdoc_path}: {str(e)}")
    
    success = len(failed) == 0
    message = f"Saved {len(saved)} mdoc files"
    if failed:
        message += f", {len(failed)} failed"
    
    return SaveAllResponse(
        success=success,
        saved=saved,
        failed=failed,
        deleted=[],
        message=message
    )


@router.post("/delete-all")
async def delete_all(request: DeleteAllRequest):
    """Delete multiple mdoc files in one request"""
    deleted = []
    failed = []
    
    for mdoc_path_str in request.mdocPaths:
        try:
            mdoc_path = Path(mdoc_path_str)
            if not mdoc_path.exists():
                failed.append(f"{mdoc_path_str}: file not found")
                continue

            # Create backup
            backup = mdoc_path.with_suffix(".mdoc.bak")
            shutil.copy2(mdoc_path, backup)

            # Delete original
            mdoc_path.unlink()

            # Remove from project state
            project_state.remove_tilt_series_by_mdoc_path(mdoc_path_str)
            
            deleted.append(mdoc_path_str)
            
        except Exception as e:
            print(f"Failed to delete {mdoc_path_str}: {e}")
            failed.append(f"{mdoc_path_str}: {str(e)}")
    
    return {
        "success": len(failed) == 0,
        "deleted": deleted,
        "failed": failed,
        "message": f"Deleted {len(deleted)} mdoc files, {len(failed)} failed"
    }

File: backend/src/routes/mdoc.rs
use std::path::Path;

use axum::{
    extract::Path as AxumPath,
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};

use crate::mdoc::parser::parse_mdoc_file;
use crate::mdoc::writer::write_mdoc_with_selections;
use crate::matcher::cut_match::ImageMatcher;
use crate::models::types::{
    BatchSaveRequest, BatchSaveResponse, BackupDeleteResponse, DeleteAllRequest,
    SaveAllRequest, SaveAllResponse, ScanConfig, TiltSeries,
};
use crate::state::project_state::PROJECT_STATE;

pub fn router() -> axum::Router {
    axum::Router::new()
        .route("/scan", axum::routing::post(scan_project))
        .route("/list", axum::routing::get(list_tilt_series))
        .route("/{ts_id}", axum::routing::get(get_tilt_series))
        .route("/save-all", axum::routing::post(save_all))
        .route("/delete-all", axum::routing::post(delete_all))
        .route("/batch-save", axum::routing::post(batch_save))
        .route("/backup-delete", axum::routing::post(backup_delete))
}

async fn scan_project(Json(config): Json<ScanConfig>) -> Result<Json<Value>, (axum::http::StatusCode, String)> {
    tracing::info!("Scanning project: {:?}", config);

    PROJECT_STATE.set_config(config.clone()).await;

    // Build image matcher cache
    let mut matcher = ImageMatcher::new(
        &config.image_dir,
        config.image_prefix_cut as usize,
        config.image_suffix_cut as usize,
    );
    matcher.build_cache();

    // Scan for mdoc files
    let mdoc_dir = Path::new(&config.mdoc_dir);
    if !mdoc_dir.exists() {
        return Err((
            axum::http::StatusCode::NOT_FOUND,
            format!("mdoc directory not found: {}", config.mdoc_dir),
        ));
    }

    let mut tilt_series: Vec<TiltSeries> = Vec::new();

    let walker = walkdir::WalkDir::new(mdoc_dir).max_depth(10);
    for entry in walker.into_iter().filter_map(|e| e.ok()) {
        if entry.file_type().is_file() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("mdoc") {
                match parse_mdoc_file(&path.to_string_lossy(), &matcher) {
                    Ok(ts) => {
                        PROJECT_STATE.add_tilt_series(ts.clone()).await;
                        tilt_series.push(ts);
                    }
                    Err(e) => {
                        tracing::warn!("Failed to parse {}: {}", path.display(), e);
                    }
                }
            }
        }
    }

    tracing::info!("Found {} tilt series", tilt_series.len());

    Ok(Json(json!({
        "tilt_series": tilt_series,
        "total": tilt_series.len()
    })))
}

async fn list_tilt_series() -> Json<Vec<TiltSeries>> {
    let series = PROJECT_STATE.list_tilt_series().await;
    Json(series)
}

async fn get_tilt_series(AxumPath(ts_id): AxumPath<String>) -> Result<Json<TiltSeries>, (axum::http::StatusCode, String)> {
    match PROJECT_STATE.get_tilt_series(&ts_id).await {
        Some(ts) => Ok(Json(ts)),
        None => Err((
            axum::http::StatusCode::NOT_FOUND,
            format!("Tilt series not found: {ts_id}"),
        )),
    }
}

async fn save_all(Json(request): Json<SaveAllRequest>) -> Json<SaveAllResponse> {
    let mut saved = Vec::new();
    let mut failed = Vec::new();

    if request.selections.is_empty() {
        return Json(SaveAllResponse {
            success: true,
            message: "No changes to save".to_string(),
            saved: vec![],
            failed: vec![],
            deleted: vec![],
        });
    }

    for (mdoc_path, selections) in &request.selections {
        match write_mdoc_with_selections(mdoc_path, selections) {
            Ok(_) => {
                PROJECT_STATE.update_tilt_series_frames(mdoc_path, selections).await;
                saved.push(mdoc_path.clone());
            }
            Err(e) => {
                tracing::error!("Failed to save {}: {}", mdoc_path, e);
                failed.push(format!("{mdoc_path}: {e}"));
            }
        }
    }

    let success = failed.is_empty();
    let message = if success {
        format!("Saved {} mdoc files", saved.len())
    } else {
        format!("Saved {} mdoc files, {} failed", saved.len(), failed.len())
    };

    Json(SaveAllResponse {
        success,
        saved,
        failed,
        deleted: vec![],
        message,
    })
}

async fn delete_all(Json(request): Json<DeleteAllRequest>) -> Json<Value> {
    let mut deleted = Vec::new();
    let mut failed = Vec::new();

    for mdoc_path_str in &request.mdoc_paths {
        let mdoc_path = Path::new(mdoc_path_str);
        if !mdoc_path.exists() {
            failed.push(format!("{mdoc_path_str}: file not found"));
            continue;
        }

        // Create backup
        let backup = mdoc_path.with_extension("mdoc.bak");
        match std::fs::copy(mdoc_path, &backup) {
            Ok(_) => {
                // Delete original
                match std::fs::remove_file(mdoc_path) {
                    Ok(_) => {
                        PROJECT_STATE.remove_tilt_series_by_mdoc_path(mdoc_path_str).await;
                        deleted.push(mdoc_path_str.clone());
                    }
                    Err(e) => {
                        failed.push(format!("{mdoc_path_str}: {e}"));
                    }
                }
            }
            Err(e) => {
                failed.push(format!("{mdoc_path_str}: backup failed - {e}"));
            }
        }
    }

    Json(json!({
        "success": failed.is_empty(),
        "deleted": deleted,
        "failed": failed,
        "message": format!("Deleted {} mdoc files, {} failed", deleted.len(), failed.len())
    }))
}

async fn batch_save(Json(request): Json<BatchSaveRequest>) -> Result<Json<BatchSaveResponse>, (axum::http::StatusCode, String)> {
    let ts = PROJECT_STATE
        .list_tilt_series()
        .await
        .into_iter()
        .find(|t| t.mdoc_path == request.mdoc_path);

    let ts = ts.ok_or_else(|| {
        (
            axum::http::StatusCode::NOT_FOUND,
            format!("Tilt series not found: {}", request.mdoc_path),
        )
    })?;

    match write_mdoc_with_selections(&request.mdoc_path, &request.selections) {
        Ok(backup_path) => {
            PROJECT_STATE
                .update_tilt_series_frames(&request.mdoc_path, &request.selections)
                .await;

            let updated_ts = PROJECT_STATE.get_tilt_series(&ts.id).await;

            Ok(Json(BatchSaveResponse {
                success: true,
                message: format!("Saved {} frame selections", request.selections.len()),
                backup_path: Some(backup_path),
                updated_tilt_series: updated_ts,
            }))
        }
        Err(e) => Err((
            axum::http::StatusCode::CONFLICT,
            format!("Save failed: {e}"),
        )),
    }
}

async fn backup_delete(Json(request): Json<BackupDeleteRequest>) -> Result<Json<BackupDeleteResponse>, (axum::http::StatusCode, String)> {
    let mdoc_path = Path::new(&request.mdoc_path);
    if !mdoc_path.exists() {
        return Err((
            axum::http::StatusCode::NOT_FOUND,
            format!("mdoc file not found: {}", request.mdoc_path),
        ));
    }

    let backup = mdoc_path.with_extension("mdoc.bak");
    std::fs::copy(mdoc_path, &backup).map_err(|e| {
        (
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            format!("Backup failed: {e}"),
        )
    })?;

    std::fs::remove_file(mdoc_path).map_err(|e| {
        (
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            format!("Delete failed: {e}"),
        )
    })?;

    PROJECT_STATE.remove_tilt_series_by_mdoc_path(&request.mdoc_path).await;

    Ok(Json(BackupDeleteResponse {
        success: true,
        message: format!("Backed up and deleted {}", request.mdoc_path),
        backup_path: Some(backup.to_string_lossy().to_string()),
    }))
}

#[derive(Deserialize)]
struct BackupDeleteRequest {
    mdoc_path: String,
}
